src/cqrs/Aggregate.ts

```diff
@@ -13,43 +13,64 @@ export class Aggregate {
     const aggregate = Aggregate.instantiateAggregate(firstEventExecutor.getAggregateType(), firstEventExecutor.getAggregateId(firstEvent), aggregateClasses);
     allEvents.forEach((e) => {
       const eventExecutor = EventExecutor.getEventExecutor(e, eventExecutors);
-      if (eventExecutor) {
-        eventExecutor.apply(e, aggregate);
-      } else {
-        throw new Error(`Unknown event during getAggregate: ${firstEvent.constructor.name}`);
-      }
+      Aggregate.applyEventOnAggregate(e, eventExecutor, aggregate);
     });
     return aggregate;
   }
 
+  public static applyEventOnAggregate(e: Object, eventExecutor: EventExecutor<any, any>, aggregate: Aggregate) {
+    if (eventExecutor) {
+      if (aggregate.eventApplied(eventExecutor.getEventId(e))) {
+        return;
+      }
+
+      eventExecutor.apply(e, aggregate);
+      const eventId = eventExecutor.getEventId(e);
+      aggregate.applyEvent(eventId);
+    } else {
+      throw new Error(`Unknown event during getEventExecutor: ${e.constructor.name}`);
+    }
+  }
+
   public static instantiateAggregate(aggregateType: string, aggregateId: string, aggregateClasses: Map<string, Function>): Aggregate {
     const aggregateClass = aggregateClasses.has(aggregateType) ? aggregateClasses.get(aggregateType) : Aggregate;
     return new (aggregateClass as any)(aggregateType, aggregateId);
   }
 
   private aggregateRoles: Map<string, Array<string>>;
+
   aggregateId: string;
   aggregateType: string;
+  private eventsCounter: number; // number of events in aggregate
+  private eventsApplied: string[];
+
   constructor(aggregateType: string, aggregateId: string) {
     this.aggregateType = aggregateType;
     this.aggregateId = aggregateId;
     this.aggregateRoles = new Map();
+    this.eventsCounter = 0;
```

Comment:
Instead of the counter please keep the maxOrdinalId this will be useful in the future when we introduce snapshots and the count of events will not have as much meaning.